pkg/controller/core/clusterqueue_controller.go

```diff
@@ -73,6 +76,27 @@ func (r *ClusterQueueReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	ctx = ctrl.LoggerInto(ctx, log)
 	log.V(2).Info("Reconciling ClusterQueue")
 
+	if cqObj.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&cqObj, kueue.ClusterQueueFinalizerName) {
+			controllerutil.AddFinalizer(&cqObj, kueue.ClusterQueueFinalizerName)
+			if err := r.client.Update(ctx, &cqObj); err != nil {
+				return ctrl.Result{}, client.IgnoreNotFound(err)
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(&cqObj, kueue.ClusterQueueFinalizerName) {
+			// The clusterQueue is being deleted, remove the finalizer only if
+			// there are no active admitted workloads.
+			if r.cache.SuspendClusterQueueIfEmpty(cqObj.Name) {
```

Comment:
We need to suspended the cluster queue immediately when it gets marked as deleted, not when it is empty. The finalizer is removed when the cluster queue becomes empty.

Also, note that in https://github.com/kubernetes-sigs/kueue/blob/b6419ef4ff9aae17c2eff2155cfb1b1fec03939d/pkg/cache/cache.go#L168 we set the cluster queue back to active whenever we get a flavor update, we need to add logic to ensure that a cq doesn't transition out of pending when it is terminating, perhaps we do need a dedicated terminating status.